In [1]:
import pandas as pd
import chainladder as cl

In [2]:
wiki = pd.read_csv('wikipedia.csv', sep=';')[['Accident Year', 'Development Year', 'Incurred']]

In [3]:
xyz_tri = cl.Triangle(
    data=wiki,
    origin="Accident Year",
    development="Development Year",
    columns=["Incurred"],
    cumulative=True,
)
xyz_tri

12          24          36          48          60          72          84          96          108         120
1998  37017487.0  43169009.0  45568919.0  46784558.0  47337318.0  47533264.0  47634419.0  47689655.0  47724678.0  47742304.0
1999  38954484.0  46045718.0  48882924.0  50219672.0  50729292.0  50926779.0  51069285.0  51163540.0  51185767.0         NaN
2000  41155776.0  49371478.0  52358476.0  53780322.0  54303086.0  54582950.0  54742188.0  54837929.0         NaN         NaN
2001  42394069.0  50584112.0  53704296.0  55150118.0  55895583.0  56156727.0  56299562.0         NaN         NaN         NaN
2002  44755243.0  52971643.0  56102312.0  57703851.0  58363564.0  58592712.0         NaN         NaN         NaN         NaN
2003  45163102.0  52497731.0  55468551.0  57015411.0  57565344.0         NaN         NaN         NaN         NaN         NaN
2004  45417309.0  52640322.0  55553673.0  56976657.0         NaN         NaN         NaN         NaN         NaN         NaN
2005  46360869.0  53790061.0  56786410.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN
2006  46582684.0  54641339.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN
2007  48853563.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN

### Age-to-age factors

In [4]:
xyz_tri.age_to_age.to_frame().style.format(precision=3)

### Simple average last 5 years

In [5]:
cl.Development(average="simple", n_periods=5).fit(xyz_tri["Incurred"]).ldf_.to_frame().style.format(precision=3)

### Simple average last 3 years

In [6]:
cl.Development(average="simple", n_periods=3).fit(xyz_tri["Incurred"]).ldf_.to_frame().style.format(precision=3)

### Volume weighted last 5 years

In [7]:
cl.Development(average="volume", n_periods=5).fit(xyz_tri["Incurred"]).ldf_.to_frame().style.format(precision=3)

### Volume weighted last 5 years

In [8]:
cl.Development(average="volume", n_periods=3).fit(xyz_tri["Incurred"]).ldf_.to_frame().style.format(precision=3)

### Cumulative to ultimate

In [9]:
cl.Development(average="volume", n_periods=3).fit(xyz_tri["Incurred"]).cdf_

12-Ult    24-Ult    36-Ult    48-Ult    60-Ult    72-Ult    84-Ult    96-Ult   108-Ult
(All)  1.289992  1.108105  1.049464  1.021522  1.009899  1.005305  1.002548  1.000949  1.000369

### Estimation of ultimate claims

In [10]:
# ???

How can I apply the cumulative development factors on the latest diagonal the get the result below?

#### Estimation of ultimate claims
|Accident year	| Reported claims	| Development factor to ultimate	| Projected ultimate claims |
|---|---|---|---|
|1998	| 47,742,304	| 1.000	| 47,742,304 |
|1999	| 51,185,767	| 1.000	| 51,185,767 |
|2000	| 54,837,929	| 1.001	| 54,892,767 |
|2001	| 56,299,562	| 1.003	| 56,468,461 |
|2002	| 58,592,712	| 1.006	| 58,944,268 |
|2003	| 57,565,344	| 1.011	| 58,198,563 | 
|2004	| 56,976,657	| 1.023	| 58,287,120 |
|2005	| 56,786,410	| 1.051	| 59,682,517 |
|2006	| 54,641,339	| 1.110	| 60,651,886 |
|2007	| 48,853,563	| 1.292	| 63,118,803 |
|Total	| 543,481,587	| 		| 569,172,456 |